### PR TITLE
Split the maybeSplit into canSplit/doSplit

### PR DESCRIPTION
**Pull request by gkdn**
This patch doesn't inline the methods to help with the review. I will follow up with another patch to inline them.

### CODE REVIEW

**Comment by gkdn** [comment]
@kripken

**Comment by gkdn** [comment]
@kripken I added another commit that inlines the functions. The first commit will be easier to review since less code is moved around.

File: src/passes/Inlining.cpp [hunk 1]
```diff
@@ -584,10 +584,12 @@ struct FunctionSplitter {
 private:
   // Information about splitting a function.
   struct Split {
-    // Whether we can split the function. If this is false, the other two will
-    // remain nullptr forever; if this is true then we will populate them the
+    enum class Kind { None, PatternA, PatternB };
```
Comment (kripken) [inline src/passes/Inlining.cpp]: Please add a comment that the patterns are described later down.

File: src/passes/Inlining.cpp [hunk 2]
```diff
@@ -607,24 +609,11 @@ struct FunctionSplitter {
   // Check if we can split a function. Returns whether we can. If the out param
   // is provided, also actually does the split, and returns the inlineable split
   // function in that out param.
-  bool maybeSplit(Function* func, Function** inlineableOut = nullptr) {
+  bool maybeSplit(Function* func) {
```
Comment (kripken) [inline src/passes/Inlining.cpp]: It looks like this function no longer does splitting, so "maybe split" might be confusing. How about `canSplit`? Looks like it returns true iff we can. Or perhaps `computeSplit` as it computes the split kind and stashes that.

File: src/passes/Inlining.cpp [hunk 3]
```diff
@@ -773,20 +737,63 @@ struct FunctionSplitter {
...
+    return true;
+  }
+
+  Function* doSplit(Function* func) {
+    auto iter = splits.find(func->name);
+    assert (iter != splits.end());
+
+    auto& split = iter->second;
+    if (split.inlineable) {
+      // We all performed the split.
```
Comment (kripken) [inline src/passes/Inlining.cpp]: `all` => `already`

File: src/passes/Inlining.cpp [hunk 4]
```diff
@@ -540,91 +540,15 @@ struct FunctionSplitter {
   // the number of iterations, etc.). Therefore this function will only find out
   // if we *can* split, but not actually do any splitting.
   bool canSplit(Function* func) {
+    // TODO: we should not get here if function marked uninlineable.
```
Comment (kripken) [inline src/passes/Inlining.cpp:543]: What did you have in mind here?

(if you mean the existing `uninlineable` field then that is only used for `canHandleParams`, which is checked below us)

File: src/passes/Inlining.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/Inlining.cpp]: Oh, sorry, it looks like the patterns appear before this actually.

File: src/passes/Inlining.cpp (hunk 2 above)
Comment (kripken) [reply src/passes/Inlining.cpp]: Looks like this was addressed in a later commit.

File: src/passes/Inlining.cpp (hunk 3 above)
Comment (kripken) [reply src/passes/Inlining.cpp]: Ah, this seems to have been removed in later commits.

File: src/passes/Inlining.cpp (hunk 4 above)
Comment (kripken) [reply src/passes/Inlining.cpp:543]: (lgtm % this question)

Comment (gkdn) [reply src/passes/Inlining.cpp:543]: I was thinking to have worthInlining to do the checking for that (had the follow up in here: cl/527177213)

Comment (kripken) [reply src/passes/Inlining.cpp:543]: I see. That sounds ok, but can you remove the comment from this PR then? (If it only makes sense in the context of the later PR, then it makes more sense there, and removing it here will avoid confusion in reading the code in the future e.g. in bisection).

**Review by kripken — APPROVED** [review]

**Pull request merged**